expression/builtin_cast_vec.go

```diff
@@ -756,11 +756,36 @@ func (b *builtinCastStringAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chu
 }
 
 func (b *builtinCastRealAsDecimalSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinCastRealAsDecimalSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETReal, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err = b.args[0].VecEvalReal(b.ctx, input, buf); err != nil {
+		return err
+	}
+	result.ResizeDecimal(n, false)
+	result.MergeNulls(buf)
+	bufreal := buf.Float64s()
+	resdecimal := result.Decimals()
+	for i := 0; i < n; i++ {
+		if !b.inUnion || bufreal[i] >= 0 {
+			if err = resdecimal[i].FromFloat64(bufreal[i]); err != nil {
+				return err
+			}
+		} else {
+			_, err = types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)
+			if err != nil {
+				return err
+			}
```

Comment:
In non-vec function, this line will run in all time whatever b.inUnion and val is (of course, without error). But in vec function (your code), it only runs when the check is false.

Reply:
I see. Is that correct?
```
for i := 0; i < n; i++ {
		if !b.inUnion || bufreal[i] >= 0 {
			if err = resdecimal[i].FromFloat64(bufreal[i]); err != nil {
				return err
			}
		}
		_, err = types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)
		if err != nil {
			return err
		}

	}
```

Reply:
resdecimal[i], err = *types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)

Reply:
I think it is ok.